### PR TITLE
fix: preserve M2M client username when no email present in role fields

### PR DESCRIPTION
**Pull request by jordane**
## Summary
- Role fields (`writers`, `auditors`, `meeting_coordinators`) can contain M2M clients that have a username but no email address
- The enrichment step in `enrichAllRoleFields()` was unconditionally clearing `username` to `""` for any entry without an email, silently destroying the M2M client's identity
- Username is now only cleared when both email and username are absent

Fixes LFXV2-2132

🤖 Generated with [Claude Code](https://claude.com/claude-code)

### CODE REVIEW

File: internal/service/project_operations.go [hunk 1]
```diff
@@ -771,7 +771,10 @@ func (s *ProjectsService) enrichAllRoleFields(
 		}
 		// Treat nil, empty, or whitespace-only emails as missing.
 		if u.Email == nil || strings.TrimSpace(*u.Email) == "" {
-			u.Username = misc.StringPtr("")
+			// M2M clients have no email but carry a valid username — preserve it.
```
[the remark this answers is not in the document]
Comment (jordane) [reply internal/service/project_operations.go]: Good catch — updated the comment to `// No email present — if a username is already set, keep it as-is.`

**Review by dealako — CHANGES_REQUESTED** [review]
Hi @jordane! 👋

This is a clean, targeted fix for a real bug — M2M clients have no email but carry a valid Auth0 client principal as their username, and the previous code silently wiped it. The logic change is minimal and correct, and the regression test is well-constructed with a realistic scenario. My main concern is a security consideration about the bypass of the `SubByEmail` identity verification path.

---

## Step 1 — Independent Review

### [minor] 🔐 Security — Username spoofing bypasses `SubByEmail` identity verification

**`internal/service/project_operations.go`, line 773–778**

The function docstring at line 750–751 explicitly states:

> _"Username misses (unknown email, empty email) write an explicit empty-string pointer so the settings converter always overwrites any previously-stored username in the database."_
> _"stale LFIDs must never be silently kept."_

This reflects an intentional design: **all identity resolution goes through `SubByEmail`; caller-supplied usernames are untrusted.** The change partially reverses that invariant: any `UserInfo` entry with a non-empty username and no email will now have its caller-supplied username preserved unmodified.

**Attack path:** An authenticated project writer calls `PUT /projects/:id/settings` with:
```json
{ "writers": [{"username": "another-user-lfid", "email": null}] }
```
The enrichment step skips `SubByEmail` (no email) and preserves `"another-user-lfid"` verbatim. The FGA sync message is then published with this relationship, granting `another-user-lfid` writer access — without email verification or that user's consent.

This requires existing writer access to exploit, which limits blast radius, but the identity verification bypass is still worth hardening. **Suggested mitigations (pick one):**
1. Extract the M2M principal from the JWT token directly rather than trusting caller-supplied usernames.
2. Validate the username format against an allow-list pattern for M2M principals (e.g., `^client-credentials\|`).
3. Add a helper that only preserves the username when the `UserInfo` has no email AND the username format matches the M2M pattern used by Auth0.

---

### [minor] 📝 Stale function-level docstring

**`internal/service/project_operations.go`, line 747–752**

Lines 750–751 and the inline comment at line 764 both still describe the old behavior ("clear username immediately for entries without an email"). Both should be updated to reflect the new invariant: _no-email entries with a pre-existing username are preserved as-is_.

Specifically:
- Line 750–751: update to note the M2M/no-email-but-has-username exception
- Line 764 (`// Gather all entries grouped by email; clear username immediately for entries without an email.`): the second half is now inaccurate

---

### [nit] 🧪 Missing edge-case test: whitespace-only username

**`internal/service/project_operations_test.go`, new test case**

The new test covers the "valid username, no email" path. The adjacent path — username is present but whitespace-only (e.g., `"   "`) — should also be cleared per the `strings.TrimSpace` guard at line 775. A small test case would pin this boundary:

```go
{
    name: "no email with whitespace-only username — username cleared",
    payload: &projsvc.UpdateProjectSettingsPayload{
        UID:     misc.StringPtr("project-uid-1"),
        IfMatch: misc.StringPtr("1"),
        Writers: []*projsvc.UserInfo{
            {Username: misc.StringPtr("   "), Name: misc.StringPtr("Whitespace User")},
        },
    },
    // mock UpdateProjectSettings: verify s.Writers[0].Username == ""
    ...
}
```

---

## Step 2 — AI Comment Reconciliation

**GitHub Copilot** ([#discussion_r3357619814](https://github.com/linuxfoundation/lfx-v2-project-service/pull/82#discussion_r3357619814)) raised that the original comment implied an M2M-only guarantee while the code preserves any non-empty username. ✅ Already addressed by the author — the comment was updated to the more accurate `"No email present — if a username is already set, keep it as-is."` No further action needed.

**CodeRabbit** summarized the change accurately and raised no issues beyond scope. The related PR #71 note is informative context.

---

## Step 3 — Revision Tracking

Single-commit PR; no prior review rounds. No regressions observed in the diff.

---

## Step 4 — Summary

🔴 **Blocking**: 0  
🟡 **Minor**: 2 — username spoofing bypass of `SubByEmail`; stale docstring  
⚪ **Nit**: 1 — whitespace-only username edge-case test

---

## 🔴 Needs changes before approval

The security concern around bypassing `SubByEmail` for caller-supplied usernames warrants a hardening pass before merge. The stale docstring (which directly documents the now-changed invariant) should also be updated. Both are straightforward fixes.

**Comment by jordane** [comment]
> The enrichment step skips `SubByEmail` (no email) and preserves `"another-user-lfid"` verbatim. The FGA sync message is then published with this relationship, granting `another-user-lfid` writer access — without email verification or that user's consent.
> 
> This requires existing writer access to exploit, which limits blast radius, but the identity verification bypass is still worth hardening. **Suggested mitigations (pick one):**
> 
> 1. Extract the M2M principal from the JWT token directly rather than trusting caller-supplied usernames.
> 2. Validate the username format against an allow-list pattern for M2M principals (e.g., `^client-credentials\|`).
> 3. Add a helper that only preserves the username when the `UserInfo` has no email AND the username format matches the M2M pattern used by Auth0.

Option 1 does not work, as the M2M client cannot call the endpoint to add permissions for themselves when they don't have permissions. It would be expected that either a user or a separate M2M client makes the call so you can't extract any info from the JWT token.

I disagree that this is a security issue. Any caller who has authenticated access to this endpoint already has writer on the project (so they don't need to add themselves), and can add any user they want -- or could register a new user and add them.

I'm not clear what is intended by option 2 vs option 3, but both seem very unnecessary here and would require us to update the codebase in the future if other username values need support, while limiting them provides no benefits.

File: internal/service/project_operations.go
Comment (andrest50) [inline internal/service/project_operations.go]: Please update the function header comments to be clear about this edge case, because right now it states that unknown emails get converted to an empty username. We should mention that because of the M2M clients which have no email, we need to not overwrite them with an empty username.

**Review by andrest50 — COMMENTED** [review]
Should we only let username be set by the request payload if it is clearly an M2M client? The idea was to not let the user of the API pass in a username to avoid fake usernames getting added, but since for M2M clients we do need to be able to provide a username then we should open it up for just those type of usernames.

**Comment by jordane** [comment]
> Should we only let username be set by the request payload if it is clearly an M2M client? The idea was to not let the user of the API pass in a username to avoid fake usernames getting added, but since for M2M clients we do need to be able to provide a username then we should open it up for just those type of usernames.

Given that our usage of clients in authelia for local development use a different pattern than auth0, and have no identifying pattern, this would be a bit tricky unless we intend to either allow list authelia clients, require authelia clients to match a pre-determined pattern, or do something else. I propose we just leave it as-is and I fixed the docstring on the function.

File: internal/service/project_operations.go
Comment (jordane) [reply internal/service/project_operations.go]: Done.

**Review by dealako — APPROVED** [review]

**Review by andrest50 — APPROVED** [review]
